### PR TITLE
QEDInternals: replace amrex::Real with amrex::ParticleReal where appropriate

### PR DESCRIPTION
**Pull request by lucafedeli88**
This PR replaces few amrex::Real with amrex::ParticleReal in the QED modules.

### CODE REVIEW

File: Source/Particles/ElementaryProcess/QEDInternals/BreitWheelerEngineWrapper.H [hunk 1]
```diff
@@ -130,10 +130,12 @@ public:
...
     int operator()(
-        const amrex::Real ux, const amrex::Real uy, const amrex::Real uz,
-        const amrex::Real ex, const amrex::Real ey, const amrex::Real ez,
-        const amrex::Real bx, const amrex::Real by, const amrex::Real bz,
-        const amrex::Real dt, amrex::Real& opt_depth) const noexcept
+        const amrex::ParticleReal ux, const amrex::ParticleReal uy,
+        const amrex::ParticleReal uz, const amrex::ParticleReal ex,
+        const amrex::ParticleReal ey, const amrex::ParticleReal ez,
+        const amrex::ParticleReal bx, const amrex::ParticleReal by,
+        const amrex::ParticleReal bz, const amrex::ParticleReal dt,
```
Comment (NeilZaim) [inline Source/Particles/ElementaryProcess/QEDInternals/BreitWheelerEngineWrapper.H]:
```suggestion
        const amrex::ParticleReal bz, const amrex::Real dt,
```

Leave `dt` as `Real`.

File: Source/Particles/ElementaryProcess/QEDInternals/BreitWheelerEngineWrapper.cpp [hunk 2]
```diff
@@ -61,7 +61,7 @@ bool BreitWheelerEngine::are_lookup_tables_initialized () const
 bool
 BreitWheelerEngine::init_lookup_tables_from_raw_data (
     const vector<char>& raw_data,
-    const amrex::Real bw_minimum_chi_phot)
+    const amrex::ParticleReal bw_minimum_chi_phot)
```
Comment (NeilZaim) [inline Source/Particles/ElementaryProcess/QEDInternals/BreitWheelerEngineWrapper.cpp]:
```suggestion
    const amrex::Real bw_minimum_chi_phot)
```

Keep this as `Real`?

File: Source/Particles/ElementaryProcess/QEDInternals/BreitWheelerEngineWrapper.cpp [hunk 3]
```diff
@@ -89,7 +89,7 @@ BreitWheelerEngine::init_lookup_tables_from_raw_data (
 }
 
 void BreitWheelerEngine::init_builtin_tables(
-    const amrex::Real bw_minimum_chi_phot)
+    const amrex::ParticleReal bw_minimum_chi_phot)
```
Comment (NeilZaim) [inline Source/Particles/ElementaryProcess/QEDInternals/BreitWheelerEngineWrapper.cpp]:
```suggestion
    const amrex::Real bw_minimum_chi_phot)
```

Keep this as `Real`?

File: Source/Particles/ElementaryProcess/QEDInternals/BreitWheelerEngineWrapper.cpp [hunk 4]
```diff
@@ -123,8 +123,8 @@ BreitWheelerEngine::get_default_ctrl() const
 {
     namespace pxr_bw = picsar::multi_physics::phys::breit_wheeler;
     return PicsarBreitWheelerCtrl{
-        pxr_bw::default_dndt_lookup_table_params<amrex::Real>,
-        pxr_bw::default_pair_prod_lookup_table_params<amrex::Real>
+        pxr_bw::default_dndt_lookup_table_params<amrex::ParticleReal>,
+        pxr_bw::default_pair_prod_lookup_table_params<amrex::ParticleReal>
```
Comment (NeilZaim) [inline Source/Particles/ElementaryProcess/QEDInternals/BreitWheelerEngineWrapper.cpp]:
```suggestion
        pxr_bw::default_dndt_lookup_table_params<amrex::Real>,
        pxr_bw::default_pair_prod_lookup_table_params<amrex::Real>
```

Keep these as reals?

File: Source/Particles/ElementaryProcess/QEDInternals/BreitWheelerEngineWrapper.cpp [hunk 5]
```diff
@@ -192,7 +192,7 @@ void BreitWheelerEngine::init_builtin_pair_prod_table()
     pair_prod_params.chi_phot_how_many = 64;
     pair_prod_params.frac_how_many = 64;
 
-    const auto vals = amrex::Gpu::DeviceVector<amrex::Real>{
+    const auto vals = amrex::Gpu::DeviceVector<amrex::ParticleReal>{
```
Comment (NeilZaim) [inline Source/Particles/ElementaryProcess/QEDInternals/BreitWheelerEngineWrapper.cpp]:
```suggestion
    const auto vals = amrex::Gpu::DeviceVector<amrex::ParticleReal>{
```

Keep this as `Real`?

File: Source/Particles/ElementaryProcess/QEDInternals/QedChiFunctions.H [hunk 6]
```diff
@@ -29,9 +29,11 @@ namespace QedUtils{
...
-        const amrex::Real ex, const amrex::Real ey, const amrex::Real ez,
-        const amrex::Real bx, const amrex::Real by, const amrex::Real bz)
+        const amrex::ParticleReal px, const amrex::ParticleReal py,
+        const amrex::ParticleReal pz, const amrex::ParticleReal ex,
+        const amrex::ParticleReal ey, const amrex::ParticleReal ez,
+        const amrex::ParticleReal bx, const amrex::ParticleReal by,
+        const amrex::ParticleReal bz)
     {
         namespace pxr_p = picsar::multi_physics::phys;
         return pxr_p::chi_photon<amrex::Real, pxr_p::unit_system::SI>(
```
Comment (NeilZaim) [inline Source/Particles/ElementaryProcess/QEDInternals/QedChiFunctions.H]:
```suggestion
        return pxr_p::chi_photon<amrex::ParticleReal, pxr_p::unit_system::SI>(
```

Use `ParticleReal` here?

File: Source/Particles/ElementaryProcess/QEDInternals/QedChiFunctions.H [hunk 7]
```diff
@@ -49,9 +51,11 @@ namespace QedUtils{
...
-        const amrex::Real ex, const amrex::Real ey, const amrex::Real ez,
-        const amrex::Real bx, const amrex::Real by, const amrex::Real bz)
+        const amrex::ParticleReal px, const amrex::ParticleReal py,
+        const amrex::ParticleReal pz, const amrex::ParticleReal ex,
+        const amrex::ParticleReal ey, const amrex::ParticleReal ez,
+        const amrex::ParticleReal bx, const amrex::ParticleReal by,
+        const amrex::ParticleReal bz)
     {
             namespace pxr_p = picsar::multi_physics::phys;
             return pxr_p::chi_ele_pos<amrex::Real, pxr_p::unit_system::SI>(
```
Comment (NeilZaim) [inline Source/Particles/ElementaryProcess/QEDInternals/QedChiFunctions.H]:
```suggestion
            return pxr_p::chi_ele_pos<amrex::ParticleReal, pxr_p::unit_system::SI>(
```

Use `ParticleReal` here?

**Review by NeilZaim — COMMENTED** [review]
Made a few suggestions but I think there are multiple ways to do this (in the end conversions between `Real` and `ParticleReal` are probably unavoidable). In my suggestions, I leave table quantities as `Real` (I guess this could be changed) and particle attributes as `ParticleReal`. That way, things should work with double precision tables and single precision particle quantities.

Other places where we could use `ParticleReal` (I didn't do an exhaustive review so there might be more places where this could be changed):
- `BreitWheelerEngineWrapper.H` line 243-245 in template argument of `pxr_m::vec3`
- `BreitWheelerEngineWrapper.H` line 248 in template argument of `generate_breit_wheeler_pairs`
- `PhotonParticleContainer.cpp` line 105 for `p_optical_depth_BW`.
- `PhysicalParticleContainer.cpp` lines 723 & 724, for `p_optical_depth_QSR` and `p_optical_depth_BW`.
- `QuantumSyncEngineWrapper.H` line 230-231 in template argument of `pxr_m::vec3`
- `QuantumSyncEngineWrapper.H` line 233 in template argument of `generate_photon_update_momentum`

**Comment by lucafedeli88** [comment]
> Made a few suggestions but I think there are multiple ways to do this (in the end conversions between `Real` and `ParticleReal` are probably unavoidable). In my suggestions, I leave table quantities as `Real` (I guess this could be changed) and particle attributes as `ParticleReal`. That way, things should work with double precision tables and single precision particle quantities.
> 
> Other places where we could use `ParticleReal` (I didn't do an exhaustive review so there might be more places where this could be changed):
> 
>     * `BreitWheelerEngineWrapper.H` line 243-245 in template argument of `pxr_m::vec3`
> 
>     * `BreitWheelerEngineWrapper.H` line 248 in template argument of `generate_breit_wheeler_pairs`
> 
>     * `PhotonParticleContainer.cpp` line 105 for `p_optical_depth_BW`.
> 
>     * `PhysicalParticleContainer.cpp` lines 723 & 724, for `p_optical_depth_QSR` and `p_optical_depth_BW`.
> 
>     * `QuantumSyncEngineWrapper.H` line 230-231 in template argument of `pxr_m::vec3`
> 
>     * `QuantumSyncEngineWrapper.H` line 233 in template argument of `generate_photon_update_momentum`

I agree with you that we should keep tables-related quantities as `Real` and that we should use `ParticleReal` where you suggested to do so

**Review by NeilZaim — APPROVED** [review]
Looks like CI is passing, so this looks good to me. Thanks for the PR.

**Review by ax3l — APPROVED** [review]

**Pull request merged**